tools for analyzing, updating and adding copyright headers in source files

isle2983 opened the pull request:
Three scripts in contrib/devtools/ are introduced:

**copyright_header_report.py**
- outputs a report of copyright claims in the source files
- helps to identify files that are missing copyright
- helps to identify files that were added that have a new copyright holder beyond the current known set
- helps to find typos in the copyright header

**copyright_header_update.py**
- replaces fix-copyright-headers.py
- does file editing in native python rather than an external call to perl

**copyright_header_insert.py**
- inserts a properly-formatted header with proper dates into a given source file

Altogether, I hope they make the task of managing the copyright headers a bit
easier.

fanquake commented:
I'm not sure that maintaining copyright headers is such an issue that it warrants splitting one tool into three. Interested in other opinions.

MarcoFalke commented:
Agree with @fanquake: I'd prefer to have the functionality (report and update) in a single script.

 Also, I don't think **copyright_header_insert.py** is required. (If someone forgets to add the header, they will forget to run the script as well)

isle2983 committed:
[devtools] script support for managing source file copyright headers

Three subcommands to this script:

1) ./copyright_header.py report

Examines git-tracked files with extensions that match:

INCLUDE = ['*.h', '*.cpp', '*.cc', '*.c', '*.py']

Helps to:

-> Identify source files without copyright
-> Identify source files added with something other than "The Bitcoin Core
developers" holder so we can be sure it is appropriate
-> Identify unintentional typos in the copyright line

2) ./copyright_header.py update

Replaces fix-copyright-headers.py. It does file editing in native python
rather than subprocessing out to perl as was the case with
fix-copyright-headers.py. It also shares code with the 'report' functions.

3) ./copyright_header.py insert

Inserts a copyright header into a source file with the proper format and
dates.

isle2983 commented:
I have condensed the three files into one and made them subcommands:
```
$ ./copyright_header_report.py <base_directory> [verbose]
$ ./copyright_header_update.py <base_directory>
$ ./copyright_header_report.py <file>
```

becomes:

```
$ ./copyright_header.py report <base_directory> [verbose]
$ ./copyright_header.py update <base_directory>
$ ./copyright_header.py insert <file>
```

running with no subcommand to copyright_header.py lists a usage string with the subcommands. Running a subcommand without arguments displays the subcommands usage string.


If we are against the 'insert' functionality, I will remove it, however I think
it is useful for a few things not mentioned:

1) It defines the exact format of a proper header
2) It gets the date right
3) It can be a building block of a strict CI system that rejects bad/missing headers and directs you to "please run` ./copyright_header.py insert <your file>`" to fix the problem.

fanquake commented:
You'll also need to update the contrib docs [here](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md#fix-copyright-headerspy).

isle2983 commented:
Thanks @fanquake. 159597a2b8967c8bc8106d98705d68932500b58a includes an update to README.md to document the new script and subcommands. Also, I tweaked some grammar and fixed some spelling in the usage strings inside copyright_header.py for clarity and made sure they matched README.md in content.

[Rendered README.md](https://github.com/isle2983/bitcoin/blob/copyright-scripts/contrib/devtools/README.md#copyright_headerpy)

laanwj commented:
Concept ACK 159597a

luke-jr commented on contrib/devtools/copyright_header.py:
```diff
@@ -0,0 +1,610 @@
+]
+EXCLUDE_COMPILED = re.compile('|'.join([fnmatch.translate(m) for m in EXCLUDE]))
+
+INCLUDE = ['*.h', '*.cpp', '*.cc', '*.c', '*.py']
```
Probably should include *.sh, *.am, *.m4, and *.include

Might make more sense as an exclusion list...

MarcoFalke replied:
We don't use `cc`

laanwj replied:
Leveldb does, though, but yes it makes no sense to include that here.

rebroad commented:
Who should run this and when?

isle2983 commented:
Good question.

I have a working branch that hooks it into TravisCI to fail branches that are missing copyright headers - forcing the submitter to either add them or to manually exclude the file from the check. I plan on submitting the PR when I get the chance to polish it a bit more

If the overall feeling is that this kind of thing is helpful and not a pain for day-to-day work, there might be utility in doing similar scripts for TravisCI to gradually ratchet in enforcement of basic whitespace rules and/or eventually clang-format and/or pylint style checks.

Also, the years in the headers need to be upgraded every so often to make sure they are current. December 31st might be a good day to run the script and post the PR.

laanwj commented:
> Who should run this and when?

- The best time to update a copyright header is when you touch the file
- The second-best is a single mass update at the end of the year of all the files that were touched in that year (that's what the scripting is for).
